libraries/persistent_store/src/bitfield.rs

```diff
@@ -0,0 +1,370 @@
+    pub fn get(&self, word: u32) -> usize {
+        ((word >> self.pos) & self.mask()) as usize
+    }
+
+    /// Sets the value of a bit field.
+    ///
+    /// # Preconditions
+    ///
+    /// - `num_bits(value) < self.len`.
+    /// - `self.get(*word) & value == value`.
+    pub fn set(&self, word: &mut u32, value: usize) {
+        let value = value as u32;
+        debug_assert_eq!(value & self.mask(), value);
+        let mask = !(self.mask() << self.pos);
+        *word &= mask | (value << self.pos);
+        debug_assert_eq!(self.get(*word), value as usize);
+    }
+
+    /// Returns a bit mask the length of the bit field.
+    fn mask(&self) -> u32 {
+        (1 << self.len) - 1
+    }
+}
+
+/// Represents a constant bit field.
+///
+/// # Invariant
+///
+/// - `num_bits(value) <= field.len`.
+pub struct Const {
+    /// The bit field.
+    pub field: Field,
+
+    /// The constant value.
+    pub value: usize,
+}
+
+impl Const {
+    /// Checks that the bit field has its value.
+    pub fn check(&self, word: u32) -> bool {
+        self.field.get(word) == self.value
+    }
+
+    /// Sets the bit field to its value.
+    pub fn set(&self, word: &mut u32) {
+        self.field.set(word, self.value);
+    }
+}
+
+/// Represents a single bit.
+///
+/// # Invariant
+///
+/// - `pos < 32`.
+pub struct Bit {
+    /// The position of the bit.
+    pub pos: usize,
+}
+
+impl Bit {
+    /// Returns whether the value of the bit is zero.
+    pub fn get(&self, word: u32) -> bool {
+        word & (1 << self.pos) == 0
+    }
+
+    /// Sets the value of the bit to zero.
+    pub fn set(&self, word: &mut u32) {
+        *word &= !(1 << self.pos);
+    }
+}
+
+/// Represents a checksum.
+///
+/// A checksum is a bit field counting how many bits are set to zero in the word (except in the
+/// checksum itself) plus some external increment. It essentially behaves like a bit field storing
+/// the external increment.
+pub struct Checksum {
+    /// The bit field
+    pub field: Field,
+}
+
+impl Checksum {
+    /// Reads the external increment from the checksum.
+    ///
+    /// # Errors
+    ///
+    /// Returns `InvalidStorage` if the external increment would be negative.
+    pub fn get(&self, word: u32) -> StoreResult<usize> {
+        let checksum = self.field.get(word);
+        let zeros = word.count_zeros() as usize - (self.field.len - checksum.count_ones() as usize);
+        checksum
+            .checked_sub(zeros)
+            .ok_or(StoreError::InvalidStorage)
+    }
+
+    /// Sets the checksum to the external increment value.
+    ///
+    /// # Preconditions
+    ///
+    /// - The checksum bits should be set to one.
+    /// - The number of zeros in the word plus the value should fit in the checksum.
+    pub fn set(&self, word: &mut u32, value: usize) {
+        debug_assert_eq!(self.field.get(*word), self.field.mask() as usize);
+        self.field.set(word, word.count_zeros() as usize + value);
```

Comment:
I hope my comment regarding `usize` versus `u32` clarified that values are natural numbers, not sequences of bits. Is the documentation about checksum still confusing? If yes, I'll try to address the remaining confusing parts.

Reply:
So

> The checksum bits should be set to one.

Actually means something like

> The bits in `word` that are covered by `field` should be set to one.

? I took me some time to dive into the whole concept, since it's a pretty bit-fiddly. But making that a little bit more verbose works for me.

Reply:
I updated the wording and added "math" version of the preconditions.